# Prediction of CO2 emissions from country-specific data
# A Machine Learning project 

***

# Stage 1: Data cleaning and preparation

***

### Notebook Contents:
0. Introduction - project and notebook summary, notes on the data source
1. Notebook setup - libraries and data import
2. Global data overview
3. Definition of the initial project goals
3. Data cleaning
    - dealing with missing values
    - transformation of the columns into a numerical data type
    - renaming of features
    - removing empty columns and rows
4. Data frame transformation
    - melting of the data for each variable
    - integration of the data into a suitable data frame format
5. Removal of missing values
    - detection of missing values
    - removal of missing values by filtering the columns and rows, so that minimal amount of features and rows are lost
5. Export the clean data frame to a file

***

## 0. Introduction

### Project summary
**Aim of the project**:
Analysis of country-specific data and development of machine learning models in order to predict CO2 emissions from country parameters. The project uses the publicly available dataset Climate Change Data from the World Bank Group, which provides data on the vast majority of countries over a range of years for parameters such as:

* country: the vast majority of countries worldwide
* year: ranging from 1990 to 2023
* CO2 emissions 
* population-specific parameter: Population density
* country economic indicators: GDP, GNI, Unemployment, etc.
* land-related parameters: Food production index, Agricultural land, and marine protected areas, Tree Cover Loss, etc.
* climate data: Nitrous oxide emissions, Cooling Degree Days, Heat Index 35 etc.
* energy use: Electricity production from coal sources, Renewable electricity output, Energy use, etc
* certain types of medical data: Life expectancy at birth, etc
* etc.



The project is divided into three stages:

1. Data cleaning and preparation
2. Data exploration and visualization
3. Predictive analysis with varios machine learning algorithms

Each of the stages is described in a separate Jupyter Notebook(.ipynp file).

***

### Notebook summary - Stage 1: Data cleaning and preparation
**Aim of this notebook**: The subject of this particular notebook is to explain the first stage of the project - the cleaning and transformation of the available data in order to prepare it for the visualization and analysis (described in further notebooks).

**Input**: comma separated values (CSV) file from the original online source

**Output**: comma separated values (CSV) file containing the cleaned data ready for visualization and analysis

**Programming language**: Python 3.8

**Libraries used in this notebook**: pandas, numpy

***

### Data source

The used data comes from the Environment, Social And Governance Data of the World Bank Group, which provides country-specific data on parameters such as CO2 emissions, energy use, Population density, Agricultural land, GDP, GNI, etc.


The dataset is publicly available at https://datacatalog.worldbank.org/search/dataset/0037651/Environment--Social-and-Governance-Data and licenced under the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International license</a>.

***


## 1. Notebook Setup
Import all needed libraries:

In [187]:
import pandas as pd
import numpy as np

In [188]:
# adjusting the path to the file is necessary, because the script is not executed in the same directory as the file
# import os
# print(os.getcwd()) # get current working directory to see where the script is executed

# Load the data into a DataFrame
file_path = 'Desktop/1.Semester/MachineLearing/Project/CO2/data/ESGEXCEL.csv' # please adjust the path to the file

try:
    data = pd.read_csv(file_path, delimiter=';', on_bad_lines='warn')
except pd.errors.ParserError as e:
    print(f"Error parsing CSV: {e}")

***

## 2. Global data overview

A global overview of the imported data 

In [189]:
print("Shape of the original dataset:")
data.shape

Shape of the original dataset:


(16969, 68)

In [190]:
# Inspect the columns of the DataFrame
print("Columns in the DataFrame:", data.columns)
print("Column data types:")
data.dtypes

Columns in the DataFrame: Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022',
       '2023'],
      dtype='object')
Column data types:


Country Name      object
Country Code      object
Indicator Name    object
Indicator Code    object
1960              object
                   ...  
2019              object
2020              object
2021              object
2022              object
2023              object
Length: 68, dtype: object

In [191]:
# Display the first few rows of the transformed DataFrame
print("Overview of the first 5 rows in the DataFrame:")
data.head()

Overview of the first 5 rows in the DataFrame:


Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name     Indicator Code 1960  \
0  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS  NaN   
1            Access to electricity (% of population)     EG.ELC.ACCS.ZS  NaN   
2  Adjusted savings: natural resources depletion ...  NY.ADJ.DRES.GN.ZS  NaN   
3  Adjusted savings: net forest depletion (% of GNI)  NY.ADJ.DFOR.GN.ZS  NaN   
4                 Agricultural land (% of land area)     AG.LND.AGRI.ZS  NaN   

         1961         1962         1963         1964         1965  ...  \
0         NaN          NaN          NaN          NaN          NaN  ...   
1         NaN          NaN          NaN          NaN          NaN  ...   
2         NaN          NaN          NaN          NaN          NaN  ...   
3         NaN          NaN          NaN          NaN          NaN  ...   
4  30,9814141  30,98266305  31,00705428  31,01800095  31,04246564  ...   

          2014         2015         2016         2017         2018  \
0  86,70571737  86,94277798  87,22870518  87,39085552  87,61786211   
1  88,83227593  89,05385159  89,53901606  90,66275374  89,17693883   
2  10,05055446  6,130654513  5,265859073  6,245422284  8,187713883   
3  0,084360637  0,096672323  0,092911395  0,102683973  0,057123056   
4   39,8344215  39,87257452  39,93781393  39,98445203  39,96973753   

          2019         2020         2021 2022 2023  
0  87,79873981  87,94826443  88,09253583  NaN  NaN  
1  90,35280229  90,63505001  90,84566129  NaN  NaN  
2  7,234435527  4,598505988          NaN  NaN  NaN  
3  0,064515779  0,075685583          NaN  NaN  NaN  
4  39,90703091  39,97329048  39,97074236  NaN  NaN  

[5 rows x 68 columns]

In [192]:
print("Descriptive statistics of the columns:")
data.describe()

Descriptive statistics of the columns:


Country Name Country Code  \
count         16969        16969   
unique          239          239   
top      Arab World          ARB   
freq             71           71   

                                           Indicator Name  Indicator Code  \
count                                               16969           16969   
unique                                                 71              71   
top     Access to clean fuels and technologies for coo...  EG.CFT.ACCS.ZS   
freq                                                  239             239   

        1960  1961  1962  1963  1964  1965  ...   2014   2015   2016   2017  \
count   1623  2156  2165  2170  2175  2212  ...  12509  12390  11783  11577   
unique  1541  2078  2090  2089  2095  2126  ...  10724  10621  10136  10025   
top        0     0     0     0     0     0  ...      0      0      0      0   
freq      27    17    17    16    15    21  ...    372    367    266    271   

         2018   2019   2020  2021  2022         2023  
count   11710  11440  10731  7215  3640          453  
unique   9972   9988   9467  6334  3543          445  
top         0      0      0   100   100  6,235160271  
freq      280    265    268   178    15            3  

[4 rows x 68 columns]

In [193]:
# get more information about the columns
data['Indicator Name'].unique() 

array(['Access to clean fuels and technologies for cooking (% of population)',
       'Access to electricity (% of population)',
       'Adjusted savings: natural resources depletion (% of GNI)',
       'Adjusted savings: net forest depletion (% of GNI)',
       'Agricultural land (% of land area)',
       'Agriculture, forestry, and fishing, value added (% of GDP)',
       'Annual freshwater withdrawals, total (% of internal resources)',
       'Annualized average growth rate in per capita real survey mean consumption or income, total population (%)',
       'Cause of death, by communicable diseases and maternal, prenatal and nutrition conditions (% of total)',
       'Children in employment, total (% of children ages 7-14)',
       'CO2 emissions (metric tons per capita)', 'Coastal protection',
       'Control of Corruption: Estimate', 'Cooling Degree Days',
       'Economic and Social Rights Performance Score',
       'Electricity production from coal sources (% of total)',
       '

In [194]:
data['Indicator Code'].unique()

array(['EG.CFT.ACCS.ZS', 'EG.ELC.ACCS.ZS', 'NY.ADJ.DRES.GN.ZS',
       'NY.ADJ.DFOR.GN.ZS', 'AG.LND.AGRI.ZS', 'NV.AGR.TOTL.ZS',
       'ER.H2O.FWTL.ZS', 'SI.SPR.PCAP.ZG', 'SH.DTH.COMM.ZS',
       'SL.TLF.0714.ZS', 'EN.ATM.CO2E.PC', 'EN.CLC.CSTP.ZS', 'CC.EST',
       'EN.CLC.CDDY.XD', 'SD.ESR.PERF.XQ', 'EG.ELC.COAL.ZS',
       'EG.IMP.CONS.ZS', 'EG.EGY.PRIM.PP.KD', 'EG.USE.PCAP.KG.OE',
       'SP.DYN.TFRT.IN', 'AG.PRD.FOOD.XD', 'AG.LND.FRST.ZS',
       'EG.USE.COMM.FO.ZS', 'NY.GDP.MKTP.KD.ZG', 'EN.CLC.GHGR.MT.CE',
       'SI.POV.GINI', 'GE.EST', 'SE.XPD.TOTL.GB.ZS', 'EN.CLC.HEAT.XD',
       'EN.CLC.HDDY.XD', 'SH.MED.BEDS.ZS', 'SI.DST.FRST.20',
       'IT.NET.USER.ZS', 'SL.TLF.ACTI.ZS', 'EN.LND.LTMP.DC',
       'ER.H2O.FWST.ZS', 'SP.DYN.LE00.IN', 'SE.ADT.LITR.ZS',
       'EN.MAM.THRD.NO', 'EN.ATM.METH.PC', 'SH.DYN.MORT', 'SM.POP.NETM',
       'EN.ATM.NOXE.PC', 'IP.PAT.RESD', 'SH.H2O.SMDW.ZS',
       'SH.STA.SMSS.ZS', 'EN.ATM.PM25.MC.M3', 'PV.EST',
       'SP.POP.65UP.TO.ZS', 'EN.POP.DNST

In [195]:
data['Country Name'].unique()

array(['Arab World', 'Caribbean small states',
       'Central Europe and the Baltics', 'Early-demographic dividend',
       'East Asia & Pacific',
       'East Asia & Pacific (excluding high income)',
       'East Asia & Pacific (IDA & IBRD)', 'Euro area',
       'Europe & Central Asia',
       'Europe & Central Asia (excluding high income)',
       'Europe & Central Asia (IDA & IBRD)', 'European Union',
       'Fragile and conflict affected situations',
       'Heavily indebted poor countries (HIPC)', 'High income',
       'IBRD only', 'IDA & IBRD total', 'IDA blend', 'IDA only',
       'IDA total', 'Late-demographic dividend',
       'Latin America & Caribbean',
       'Latin America & Caribbean (excluding high income)',
       'Latin America & Caribbean (IDA & IBRD)',
       'Least developed countries: UN classification',
       'Low & middle income', 'Low income', 'Lower middle income',
       'Middle East & North Africa',
       'Middle East & North Africa (excluding high income)

### Findings from the global overview**

This global overview gives away the following facts about the available data:

- shape: 68 columns, 16969 rows
- all columns are of type "object" - neither numeric, nor string/text values
A certain amount of missing values, denoted both as NaN (not a number values)
- The columns represent key values such as country, but also the corresponding years and the indicator code/name
- The columns 'Country Code' and 'Indicator Code' do not give any information and are therefore obsolete
- The column 'Indicator Name' contains the country-specific features required for the analysis
- The names of the features in the column 'Indicator Name' are clear but too long

*** 

## 3. Define the initial project goals

The first overview of the raw data allows to define initial goals and objectives of the machine learning project. These will be refined in the future as more information insight is gained from the data. However, this initial goal definition will help develop a strategy and organize the data cleaning, transformation and visualization.

The data series available can be summarized into the following country-specific parameter/feature categories:

* target variable: CO2 emissions 

Features:

* country: the vast majority of countries worldwide
* year: ranging from 1990 to 2023
* population-specific parameter: Population density
* country economic indicators: GDP, GNI, Unemployment, etc.
* land-related parameters: Food production index, Agricultural land, and marine protected areas, Tree Cover Loss, etc.
* climate data: Nitrous oxide emissions, Cooling Degree Days, Heat Index 35 etc.
* energy use: Electricity production from coal sources, Renewable electricity output, Energy use, etc
* certain types of medical data: Life expectancy at birth, etc
* etc.

Such a dataset would suggest to investigate the influence of country-specific parameters such as economic parameres, population, energy use, land use and others on climate-related data or the factors affecting the climate like emissions, precipitations, etc.

**Initial goal of the machine learning project:** Analyze the relationships among these variable categories and evaluate the contribution of factors like country economy, energy use, land use, etc. on greenhouse gas emissions, precipitations, etc. Finally, develop a machine learning model capable of predicting climate-related data or emissions from the other country-specific parameters.

As more data insight will be gained with along the course of the project, the definition of these goals will be refined in more detail.

***

## 4. Data cleaning

### Organization of the data cleaning and transformation

The main aim of the data cleaning and transformation is to represent the features (the country parameters contained in the column *'Indicator Name'*) as separate columns and to make each row identifiable by a country and a year. At the same time, it would make sense to transform the years into a single column.

Additionally, it is necessary to get rid of empty rows or columns and deal with the remaining cells with missing values.

For these purposes, the following tasks have to be undertaken:

1. Remove the unnecessary columns "Country Code", "Indicator Code"
2. Transform the year columns into a numerical data type
3. Rename the features in column "Indicator Name"

### 4.1 Removing the unnecessary columns "Country Code", "Indicator Code"

In [196]:
# assign the data to a new DataFrame, which will be modified
data_clean = data

print("Original number of columns:")
print(data_clean.shape[1])

data_clean = data_clean.drop(['Country Code', 'Indicator Code'], axis='columns')

print("Current number of columns:")
print(data_clean.shape[1])

Original number of columns:
68
Current number of columns:
66


### 4.2 Transform the year columns into a numerical data type

In [197]:
data_clean.head()

Country Name                                     Indicator Name 1960  \
0   Arab World  Access to clean fuels and technologies for coo...  NaN   
1   Arab World            Access to electricity (% of population)  NaN   
2   Arab World  Adjusted savings: natural resources depletion ...  NaN   
3   Arab World  Adjusted savings: net forest depletion (% of GNI)  NaN   
4   Arab World                 Agricultural land (% of land area)  NaN   

         1961         1962         1963         1964         1965  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4  30,9814141  30,98266305  31,00705428  31,01800095  31,04246564   

          1966         1967  ...         2014         2015         2016  \
0          NaN          NaN  ...  86,70571737  86,94277798  87,22870518   
1          NaN          NaN  ...  88,83227593  89,05385159  89,53901606   
2          NaN          NaN  ...  10,05055446  6,130654513  5,265859073   
3          NaN          NaN  ...  0,084360637  0,096672323  0,092911395   
4  31,05040014  31,10322331  ...   39,8344215  39,87257452  39,93781393   

          2017         2018         2019         2020         2021 2022 2023  
0  87,39085552  87,61786211  87,79873981  87,94826443  88,09253583  NaN  NaN  
1  90,66275374  89,17693883  90,35280229  90,63505001  90,84566129  NaN  NaN  
2  6,245422284  8,187713883  7,234435527  4,598505988          NaN  NaN  NaN  
3  0,102683973  0,057123056  0,064515779  0,075685583          NaN  NaN  NaN  
4  39,98445203  39,96973753  39,90703091  39,97329048  39,97074236  NaN  NaN  

[5 rows x 66 columns]

In [198]:
print(data_clean.dtypes)

Country Name      object
Indicator Name    object
1960              object
1961              object
1962              object
                   ...  
2019              object
2020              object
2021              object
2022              object
2023              object
Length: 66, dtype: object


In [199]:
# Function to clean numeric columns starting from the second column
def clean_numeric_columns(df):
    for col in df.columns[2:]:  # Skip the first two columns
        print(f"Converting column {col} to numeric")
        
        # Remove extra whitespace, replace commas with dots, and convert to numeric
        df[col] = df[col].str.strip().str.replace(',', '.', regex=True)  # Make sure to assign the result back to the column
        
        # Convert to numeric
        df[col] = pd.to_numeric(df[col], errors='coerce')  # Coerce invalid values to NaN to avoid ValueError
        
        print(f"Converted column {col} to numeric")
    return df

# Apply the function to your dataset
data_clean2 = clean_numeric_columns(data_clean.copy())


Converting column 1960 to numeric
Converted column 1960 to numeric
Converting column 1961 to numeric
Converted column 1961 to numeric
Converting column 1962 to numeric
Converted column 1962 to numeric
Converting column 1963 to numeric
Converted column 1963 to numeric
Converting column 1964 to numeric
Converted column 1964 to numeric
Converting column 1965 to numeric


Converted column 1965 to numeric
Converting column 1966 to numeric
Converted column 1966 to numeric
Converting column 1967 to numeric
Converted column 1967 to numeric
Converting column 1968 to numeric
Converted column 1968 to numeric
Converting column 1969 to numeric
Converted column 1969 to numeric
Converting column 1970 to numeric
Converted column 1970 to numeric
Converting column 1971 to numeric
Converted column 1971 to numeric
Converting column 1972 to numeric
Converted column 1972 to numeric
Converting column 1973 to numeric
Converted column 1973 to numeric
Converting column 1974 to numeric
Converted column 1974 to numeric
Converting column 1975 to numeric
Converted column 1975 to numeric
Converting column 1976 to numeric
Converted column 1976 to numeric
Converting column 1977 to numeric
Converted column 1977 to numeric
Converting column 1978 to numeric
Converted column 1978 to numeric
Converting column 1979 to numeric
Converted column 1979 to numeric
Converting column 1980 to num

In [200]:
data_clean2.head()

Country Name                                     Indicator Name  1960  \
0   Arab World  Access to clean fuels and technologies for coo...   NaN   
1   Arab World            Access to electricity (% of population)   NaN   
2   Arab World  Adjusted savings: natural resources depletion ...   NaN   
3   Arab World  Adjusted savings: net forest depletion (% of GNI)   NaN   
4   Arab World                 Agricultural land (% of land area)   NaN   

        1961       1962       1963       1964       1965     1966       1967  \
0        NaN        NaN        NaN        NaN        NaN      NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN      NaN        NaN   
2        NaN        NaN        NaN        NaN        NaN      NaN        NaN   
3        NaN        NaN        NaN        NaN        NaN      NaN        NaN   
4  30.981414  30.982663  31.007054  31.018001  31.042466  31.0504  31.103223   

   ...       2014       2015       2016       2017       2018       2019  \
0  ...  86.705717  86.942778  87.228705  87.390856  87.617862  87.798740   
1  ...  88.832276  89.053852  89.539016  90.662754  89.176939  90.352802   
2  ...  10.050554   6.130655   5.265859   6.245422   8.187714   7.234436   
3  ...   0.084361   0.096672   0.092911   0.102684   0.057123   0.064516   
4  ...  39.834421  39.872575  39.937814  39.984452  39.969738  39.907031   

        2020       2021  2022  2023  
0  87.948264  88.092536   NaN   NaN  
1  90.635050  90.845661   NaN   NaN  
2   4.598506        NaN   NaN   NaN  
3   0.075686        NaN   NaN   NaN  
4  39.973290  39.970742   NaN   NaN  

[5 rows x 66 columns]

In [201]:
data_clean2.dtypes

Country Name       object
Indicator Name     object
1960              float64
1961              float64
1962              float64
                   ...   
2019              float64
2020              float64
2021              float64
2022              float64
2023              float64
Length: 66, dtype: object

In [202]:
# Filter the DataFrame to include only rows where 'Indicator Name' is 'CO2 emissions (metric tons per capita)'
# and then drop rows with NaN values in the 'CO2 emissions (metric tons per capita)' column

filtered_df = data[(data['Indicator Name'] == 'CO2 emissions (metric tons per capita)')]

# Select the 'country' and 'CO2 emissions (metric tons per capita)' columns, drop duplicates, and show the first 100 rows
unique_co2_emissions = filtered_df[['Country Name', 'Indicator Name', '1991']].drop_duplicates().head(100)

unique_co2_emissions.notnull().sum()


Country Name      100
Indicator Name    100
1991              100
dtype: int64

***

## 5. Data frame transformation

This is how the current data frame looks like:

In [203]:
data.head(100)

Country Name Country Code  \
0               Arab World          ARB   
1               Arab World          ARB   
2               Arab World          ARB   
3               Arab World          ARB   
4               Arab World          ARB   
..                     ...          ...   
95  Caribbean small states          CSS   
96  Caribbean small states          CSS   
97  Caribbean small states          CSS   
98  Caribbean small states          CSS   
99  Caribbean small states          CSS   

                                       Indicator Name     Indicator Code 1960  \
0   Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS  NaN   
1             Access to electricity (% of population)     EG.ELC.ACCS.ZS  NaN   
2   Adjusted savings: natural resources depletion ...  NY.ADJ.DRES.GN.ZS  NaN   
3   Adjusted savings: net forest depletion (% of GNI)  NY.ADJ.DFOR.GN.ZS  NaN   
4                  Agricultural land (% of land area)     AG.LND.AGRI.ZS  NaN   
..                                                ...                ...  ...   
95  GHG net emissions/removals by LUCF (Mt of CO2 ...  EN.CLC.GHGR.MT.CE  NaN   
96                                         Gini index        SI.POV.GINI  NaN   
97                 Government Effectiveness: Estimate             GE.EST  NaN   
98  Government expenditure on education, total (% ...  SE.XPD.TOTL.GB.ZS  NaN   
99                                      Heat Index 35     EN.CLC.HEAT.XD  NaN   

          1961         1962         1963         1964         1965  ...  \
0          NaN          NaN          NaN          NaN          NaN  ...   
1          NaN          NaN          NaN          NaN          NaN  ...   
2          NaN          NaN          NaN          NaN          NaN  ...   
3          NaN          NaN          NaN          NaN          NaN  ...   
4   30,9814141  30,98266305  31,00705428  31,01800095  31,04246564  ...   
..         ...          ...          ...          ...          ...  ...   
95         NaN          NaN          NaN          NaN          NaN  ...   
96         NaN          NaN          NaN          NaN          NaN  ...   
97         NaN          NaN          NaN          NaN          NaN  ...   
98         NaN          NaN          NaN          NaN          NaN  ...   
99         NaN          NaN          NaN          NaN          NaN  ...   

           2014         2015         2016         2017         2018  \
0   86,70571737  86,94277798  87,22870518  87,39085552  87,61786211   
1   88,83227593  89,05385159  89,53901606  90,66275374  89,17693883   
2   10,05055446  6,130654513  5,265859073  6,245422284  8,187713883   
3   0,084360637  0,096672323  0,092911395  0,102683973  0,057123056   
4    39,8344215  39,87257452  39,93781393  39,98445203  39,96973753   
..          ...          ...          ...          ...          ...   
95          NaN          NaN          NaN          NaN          NaN   
96          NaN          NaN          NaN          NaN          NaN   
97          NaN          NaN          NaN          NaN          NaN   
98   15,1567359  14,37622547  14,78958893  14,14472294  14,17787075   
99          NaN          NaN          NaN          NaN          NaN   

           2019         2020         2021         2022 2023  
0   87,79873981  87,94826443  88,09253583          NaN  NaN  
1   90,35280229  90,63505001  90,84566129          NaN  NaN  
2   7,234435527  4,598505988          NaN          NaN  NaN  
3   0,064515779  0,075685583          NaN          NaN  NaN  
4   39,90703091  39,97329048  39,97074236          NaN  NaN  
..          ...          ...          ...          ...  ...  
95          NaN          NaN          NaN          NaN  NaN  
96          NaN          NaN          NaN          NaN  NaN  
97          NaN          NaN          NaN          NaN  NaN  
98   13,9673872  11,71627045  13,72220993  8,589530945  NaN  
99          NaN          NaN          NaN          NaN  NaN  

[100 rows x 68 columns]

In [205]:
# Define shorter names corresponding to most relevant variables in a dictionary
chosen_vars = {
    'Access to clean fuels and technologies for cooking (% of population)': 'clean_fuel_access_perc',
    'Access to electricity (% of population)': 'elec_access_perc',
    'Adjusted savings: natural resources depletion (% of GNI)': 'nat_res_depl_perc',
    'Adjusted savings: net forest depletion (% of GNI)': 'forest_depl_perc',
    'Agricultural land (% of land area)': 'agri_land_perc',
    'Agriculture, forestry, and fishing, value added (% of GDP)': 'agri_forest_fish_val_perc',
    'Annual freshwater withdrawals, total (% of internal resources)': 'freshwater_withdrawal_perc',
    'Annualized average growth rate in per capita real survey mean consumption or income, total population (%)': 'avg_growth_rate_consumption_income_perc',
    'Cause of death, by communicable diseases and maternal, prenatal and nutrition conditions (% of total)': 'death_communicable_diseases_perc',
    'Children in employment, total (% of children ages 7-14)': 'children_employment_perc',
    'CO2 emissions (metric tons per capita)': 'co2_emissions_per_capita',
    'Coastal protection': 'coastal_protection',
    'Control of Corruption: Estimate': 'corruption_control_est',
    'Cooling Degree Days': 'cooling_degree_days',
    'Economic and Social Rights Performance Score': 'econ_social_rights_score',
    'Electricity production from coal sources (% of total)': 'elec_prod_coal_perc',
    'Energy imports, net (% of energy use)': 'energy_imports_net_perc',
    'Energy intensity level of primary energy (MJ/$2017 PPP GDP)': 'energy_intensity_primary_energy',
    'Energy use (kg of oil equivalent per capita)': 'energy_use_per_capita',
    'Fertility rate, total (births per woman)': 'fertility_rate',
    'Food production index (2014-2016 = 100)': 'food_prod_index',
    'Forest area (% of land area)': 'forest_area_perc',
    'Fossil fuel energy consumption (% of total)': 'fossil_fuel_energy_consumption_perc',
    'GDP growth (annual %)': 'gdp_growth_perc',
    'GHG net emissions/removals by LUCF (Mt of CO2 equivalent)': 'ghg_net_emissions_lucf',
    'Gini index': 'gini_index',
    'Government Effectiveness: Estimate': 'gov_effectiveness_est',
    'Government expenditure on education, total (% of government expenditure)': 'gov_expenditure_education_perc',
    'Heat Index 35': 'heat_index_35',
    'Heating Degree Days': 'heating_degree_days',
    'Hospital beds (per 1,000 people)': 'hospital_beds_per_1000',
    'Income share held by lowest 20%': 'income_share_lowest_20',
    'Individuals using the Internet (% of population)': 'internet_usage_perc',
    'Labor force participation rate, total (% of total population ages 15-64) (modeled ILO estimate)': 'labor_force_participation_perc',
    'Land Surface Temperature': 'land_surface_temp',
    'Level of water stress: freshwater withdrawal as a proportion of available freshwater resources': 'water_stress_level',
    'Life expectancy at birth, total (years)': 'life_expectancy',
    'Literacy rate, adult total (% of people ages 15 and above)': 'literacy_rate_adult',
    'Mammal species, threatened': 'mammal_species_threatened',
    'Methane emissions (metric tons of CO2 equivalent per capita)': 'methane_emissions_per_capita',
    'Mortality rate, under-5 (per 1,000 live births)': 'mortality_rate_under_5',
    'Net migration': 'net_migration',
    'Nitrous oxide emissions (metric tons of CO2 equivalent per capita)': 'nitrous_oxide_emissions_per_capita',
    'Patent applications, residents': 'patent_applications_residents',
    'People using safely managed drinking water services (% of population)': 'safe_drinking_water_perc',
    'People using safely managed sanitation services (% of population)': 'safe_sanitation_services_perc',
    'PM2.5 air pollution, mean annual exposure (micrograms per cubic meter)': 'pm25_air_pollution',
    'Political Stability and Absence of Violence/Terrorism: Estimate': 'political_stability_est',
    'Population ages 65 and above (% of total population)': 'population_65_above_perc',
    'Population density (people per sq. km of land area)': 'population_density',
    'Poverty headcount ratio at national poverty lines (% of population)': 'poverty_headcount_ratio',
    'Prevalence of overweight (% of adults)': 'prevalence_overweight',
    'Prevalence of undernourishment (% of population)': 'prevalence_undernourishment',
    'Proportion of bodies of water with good ambient water quality': 'water_quality_good_perc',
    'Proportion of seats held by women in national parliaments (%)': 'women_parliament_seats_perc',
    'Ratio of female to male labor force participation rate (%) (modeled ILO estimate)': 'female_to_male_labor_force_ratio',
    'Regulatory Quality: Estimate': 'regulatory_quality_est',
    'Renewable electricity output (% of total electricity output)': 'renewable_elec_output_perc',
    'Renewable energy consumption (% of total final energy consumption)': 'renewable_energy_consumption_perc',
    'Research and development expenditure (% of GDP)': 'rnd_expenditure_perc',
    'Rule of Law: Estimate': 'rule_of_law_est',
    'School enrollment, primary (% gross)': 'school_enrollment_primary_perc',
    'School enrollment, primary and secondary (gross), gender parity index (GPI)': 'school_enrollment_gender_parity',
    'Scientific and technical journal articles': 'scientific_journal_articles',
    'Standardised Precipitation-Evapotranspiration Index': 'spei',
    'Strength of legal rights index (0=weak to 12=strong)': 'legal_rights_index',
    'Terrestrial and marine protected areas (% of total territorial area)': 'protected_areas_perc',
    'Tree Cover Loss (hectares)': 'tree_cover_loss',
    'Unemployment, total (% of total labor force) (modeled ILO estimate)': 'unemployment_total_perc',
    'Unmet need for contraception (% of married women ages 15-49)': 'unmet_need_contraception',
    'Voice and Accountability: Estimate': 'voice_accountability_est'
}

data_clean['Indicator Name'] = data_clean['Indicator Name'].replace(to_replace=chosen_vars)

data_clean.head()


Country Name          Indicator Name 1960        1961         1962  \
0   Arab World  clean_fuel_access_perc  NaN         NaN          NaN   
1   Arab World        elec_access_perc  NaN         NaN          NaN   
2   Arab World       nat_res_depl_perc  NaN         NaN          NaN   
3   Arab World        forest_depl_perc  NaN         NaN          NaN   
4   Arab World          agri_land_perc  NaN  30,9814141  30,98266305   

          1963         1964         1965         1966         1967  ...  \
0          NaN          NaN          NaN          NaN          NaN  ...   
1          NaN          NaN          NaN          NaN          NaN  ...   
2          NaN          NaN          NaN          NaN          NaN  ...   
3          NaN          NaN          NaN          NaN          NaN  ...   
4  31,00705428  31,01800095  31,04246564  31,05040014  31,10322331  ...   

          2014         2015         2016         2017         2018  \
0  86,70571737  86,94277798  87,22870518  87,39085552  87,61786211   
1  88,83227593  89,05385159  89,53901606  90,66275374  89,17693883   
2  10,05055446  6,130654513  5,265859073  6,245422284  8,187713883   
3  0,084360637  0,096672323  0,092911395  0,102683973  0,057123056   
4   39,8344215  39,87257452  39,93781393  39,98445203  39,96973753   

          2019         2020         2021 2022 2023  
0  87,79873981  87,94826443  88,09253583  NaN  NaN  
1  90,35280229  90,63505001  90,84566129  NaN  NaN  
2  7,234435527  4,598505988          NaN  NaN  NaN  
3  0,064515779  0,075685583          NaN  NaN  NaN  
4  39,90703091  39,97329048  39,97074236  NaN  NaN  

[5 rows x 66 columns]

In [206]:
import pandas as pd
from functools import reduce

# Save the feature names into a list of strings
chosen_cols = data_clean['Indicator Name'].unique()

# Define an empty list, where sub-dataframes for each feature will be saved
frame_list = []

# Iterate over all chosen features
for variable in chosen_cols:
    if variable in chosen_vars:
        # Pick only rows corresponding to the current feature
        frame = data_clean[data_clean['Indicator Name'] == variable]
        
        # Melt all the values for all years into one column and rename the columns correspondingly
        frame = frame.melt(id_vars=['Country Name', 'Indicator Name'], var_name='year', value_name=chosen_vars[variable])
        frame = frame.rename(columns={'Country Name': 'country'}).drop(['Indicator Name'], axis='columns')
        
        # Add the melted dataframe for the current feature into the list
        frame_list.append(frame)

        # Debugging: Print the intermediate dataframe
        print(f"DataFrame for {variable}:\n", frame.head())

# Check if frame_list is not empty before merging
if frame_list:
    # Merge all sub-frames into a single dataframe, making an outer binding on the key columns 'country','year'
    all_vars = reduce(lambda left, right: pd.merge(left, right, on=['country', 'year'], how='outer'), frame_list)
else:
    # Create an empty dataframe with the expected columns if frame_list is empty
    all_vars = pd.DataFrame(columns=['country', 'year'] + list(chosen_vars.values()))

# Debugging: Print the final merged DataFrame
print("Final merged DataFrame:\n", all_vars.head())

Final merged DataFrame:
 Empty DataFrame
Columns: [country, year, clean_fuel_access_perc, elec_access_perc, nat_res_depl_perc, forest_depl_perc, agri_land_perc, agri_forest_fish_val_perc, freshwater_withdrawal_perc, avg_growth_rate_consumption_income_perc, death_communicable_diseases_perc, children_employment_perc, co2_emissions_per_capita, coastal_protection, corruption_control_est, cooling_degree_days, econ_social_rights_score, elec_prod_coal_perc, energy_imports_net_perc, energy_intensity_primary_energy, energy_use_per_capita, fertility_rate, food_prod_index, forest_area_perc, fossil_fuel_energy_consumption_perc, gdp_growth_perc, ghg_net_emissions_lucf, gini_index, gov_effectiveness_est, gov_expenditure_education_perc, heat_index_35, heating_degree_days, hospital_beds_per_1000, income_share_lowest_20, internet_usage_perc, labor_force_participation_perc, land_surface_temp, water_stress_level, life_expectancy, literacy_rate_adult, mammal_species_threatened, methane_emissions_per_capit

In [208]:
all_vars['co2_emissions_per_capita'].notnull().sum()

0

In [209]:
all_vars.head().notnull().sum() 

country                     0
year                        0
clean_fuel_access_perc      0
elec_access_perc            0
nat_res_depl_perc           0
                           ..
protected_areas_perc        0
tree_cover_loss             0
unemployment_total_perc     0
unmet_need_contraception    0
voice_accountability_est    0
Length: 73, dtype: int64

***

## 6. Remove the remaining missing values in an optimal way

Although some columns and rows with empty cells have already been deleted, there are still remaining missing values:

In [210]:
print("check the amount of missing values in each column")
all_vars.isnull().sum()

check the amount of missing values in each column


country                     0
year                        0
clean_fuel_access_perc      0
elec_access_perc            0
nat_res_depl_perc           0
                           ..
protected_areas_perc        0
tree_cover_loss             0
unemployment_total_perc     0
unmet_need_contraception    0
voice_accountability_est    0
Length: 73, dtype: int64

### Filtering the countries and removing coutries without the target variable


In [211]:
# Select the 'Country Name' and 'CO2 emissions (metric tons per capita)' columns, drop duplicates
unique_co2_emissions = all_vars[['country', 'CO2 emissions (metric tons per capita)']].drop_duplicates()

# Print all the countries and their CO2 emissions
print(unique_co2_emissions.to_string(index=False))

KeyError: "['CO2 emissions (metric tons per capita)'] not in index"

In [48]:
all_vars_clean = all_vars

# delete rows with missing values in the column 'CO2 emissions (metric tons per capita)'
all_vars_clean = all_vars_clean.dropna(subset=['CO2 emissions (metric tons per capita)'])

In [49]:
# get all countries 
all_vars_clean['country'].unique()

array(['Eritrea', 'Marshall Islands', 'Micronesia, Fed. Sts.', 'Palau',
       'Timor-Leste'], dtype=object)

In [50]:
# get all countreis with missing CO2 emissions

all_vars_clean.head()

country  year  \
3934                Eritrea  1990   
3935                Eritrea  1991   
8734       Marshall Islands  1990   
8735       Marshall Islands  1991   
8990  Micronesia, Fed. Sts.  1990   

      Access to clean fuels and technologies for cooking (% of population)  \
3934                                                NaN                      
3935                                                NaN                      
8734                                                NaN                      
8735                                                NaN                      
8990                                                NaN                      

      Access to electricity (% of population)  \
3934                                      NaN   
3935                                      NaN   
8734                                      NaN   
8735                                      NaN   
8990                                      NaN   

      Adjusted savings: natural resources depletion (% of GNI)  \
3934                                                NaN          
3935                                                NaN          
8734                                                NaN          
8735                                                NaN          
8990                                                NaN          

      Adjusted savings: net forest depletion (% of GNI)  \
3934                                                NaN   
3935                                                NaN   
8734                                                NaN   
8735                                                NaN   
8990                                                NaN   

      Agricultural land (% of land area)  \
3934                                 NaN   
3935                                 NaN   
8734                                 NaN   
8735                                 NaN   
8990                                 NaN   

      Agriculture, forestry, and fishing, value added (% of GDP)  \
3934                                                NaN            
3935                                                NaN            
8734                                                NaN            
8735                                                NaN            
8990                                                NaN            

      Annual freshwater withdrawals, total (% of internal resources)  \
3934                                                NaN                
3935                                                NaN                
8734                                                NaN                
8735                                                NaN                
8990                                                NaN                

      Annualized average growth rate in per capita real survey mean consumption or income, total population (%)  \
3934                                                NaN                                                           
3935                                                NaN                                                           
8734                                                NaN                                                           
8735                                                NaN                                                           
8990                                                NaN                                                           

      ...  School enrollment, primary (% gross)  \
3934  ...                                   NaN   
3935  ...                                   NaN   
8734  ...                                   NaN   
8735  ...                                   NaN   
8990  ...                                   NaN   

      School enrollment, primary and secondary (gross), gender parity index (GPI)  \
3934                                                NaN                             
3935  

### 6.1 Filtering the years by missing values

Checking the amount of missing values for each year:

In [51]:
#define an array with the unique year values
years_count_missing = dict.fromkeys(all_vars_clean['year'].unique(), 0)
for ind, row in all_vars_clean.iterrows():
    years_count_missing[row['year']] += row.isnull().sum()

# sort the years by missing values
years_missing_sorted = dict(sorted(years_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each year
print("missing values by year:")
for key, val in years_missing_sorted.items():
    print(key, ":", val)

missing values by year:
1992 : 64
1993 : 64
1995 : 64
1996 : 64
1997 : 64
1998 : 64
1999 : 64
1994 : 65
2001 : 65
2000 : 66
1990 : 321
1991 : 322


The purpose of the filtering is to delete rows with a significant amount of missing values for certain countries without removing too many years. So it is important to choose the proper limit for NaN values allowed per year. The previous output suggests to pick the years between 1992 and 2023 for the further analysis (since they are all below 16.00 NaN values):

In [52]:
all_vars_clean.head()
# get type of the year column
# all_vars_clean['year'].dtype
# how many unique countries are there in the dataset
all_vars_clean['country'].unique()

# transform the year column to integer
# all_vars_clean['year'] = all_vars_clean['year'].astype(int)

array(['Eritrea', 'Marshall Islands', 'Micronesia, Fed. Sts.', 'Palau',
       'Timor-Leste'], dtype=object)

In [53]:
print("number of missing values in the whole dataset before filtering the years:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows before filtering the years:")
print(all_vars_clean.shape[0])

# filter only rows for years between 1991 and 2008 (having less missing values)
all_vars_clean = all_vars_clean[(all_vars_clean['year'] >= 1992) & (all_vars_clean['year'] <= 2023)]

print("number of missing values in the whole dataset after filtering the years:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows after filtering the years:")
print(all_vars_clean.shape[0])

number of missing values in the whole dataset before filtering the years:
1287
number of rows before filtering the years:
20


TypeError: '>=' not supported between instances of 'str' and 'int'

### 6.2 Filtering the countries by missing values

The same procedure is applied to the filtering of countries with missing values. The following snippet shows the number of NaNs for each country.

In [66]:
# check the amount of missing values by country

# define an array with the unique country values
countries_count_missing = dict.fromkeys(all_vars_clean['country'].unique(), 0)

# iterate through all rows and count the amount of NaN values for each country
for ind, row in all_vars_clean.iterrows():
    countries_count_missing[row['country']] += row.isnull().sum()

# sort the countries by missing values
countries_missing_sorted = dict(sorted(countries_count_missing.items(), key=lambda item: item[1]))

# print the missing values for each country
print("missing values by country:")
for key, val in countries_missing_sorted.items():
    print(key, ":", val)

missing values by country:
Timor-Leste : 644


This output would suggest to remove rows for countries with more than 1.300 missing values since we are also getting rid of the columns where countries are combinded. 

In [67]:
print("number of missing values in the whole dataset before filtering the countries:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows before filtering the countries:")
print(all_vars_clean.shape[0])


# filter only rows for countries with less than 90 missing values
countries_filter = []
for key, val in countries_missing_sorted.items():
    if val<1300:
        countries_filter.append(key)

all_vars_clean = all_vars_clean[all_vars_clean['country'].isin(countries_filter)]

print("number of missing values in the whole dataset after filtering the countries:")
print(all_vars_clean.isnull().sum().sum())
print("number of rows after filtering the countries:")
print(all_vars_clean.shape[0])

number of missing values in the whole dataset before filtering the countries:
644
number of rows before filtering the countries:
10
number of missing values in the whole dataset after filtering the countries:
644
number of rows after filtering the countries:
10


### 6.3 Checking the features (columns) for missing values

The NaN values count in each column is:

In [68]:
all_vars_clean.isnull().sum().sort_values(ascending=False)

Land Surface Temperature                                                                                     10
Poverty headcount ratio at national poverty lines (% of population)                                          10
Level of water stress: freshwater withdrawal as a proportion of available freshwater resources               10
Life expectancy at birth, total (years)                                                                      10
Literacy rate, adult total (% of people ages 15 and above)                                                   10
Mammal species, threatened                                                                                   10
Methane emissions (metric tons of CO2 equivalent per capita)                                                 10
Mortality rate, under-5 (per 1,000 live births)                                                              10
Nitrous oxide emissions (metric tons of CO2 equivalent per capita)                                      